catman.8
.\" $Id$
.\"
.\" Copyright (c) 2017, 2025 Ingo Schwarze <schwarze@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt CATMAN 8
.Os
.Sh NAME
.Nm catman
.Nd format all manual pages below a directory
.Sh SYNOPSIS
.Nm catman
.Op Fl I Cm os Ns = Ns Ar name
.Op Fl T Ar output
.Ar srcdir dstdir
.Sh DESCRIPTION
The
.Nm
utility assumes that all files below
.Ar srcdir
are manual pages in
.Xr  mdoc 7
and
.Xr  man 7
format and formats all of them, storing the formatted versions in
the same relative paths below
.Ar dstdir .
Unless they already exist,
.Ar dstdir
itself and any required subdirectories are created.
Existing files are not explicitly deleted, but possibly overwritten.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl I Cm os Ns = Ns Ar name
Override the default operating system
.Ar name
for the
.Xr  mdoc 7
.Ic \&Os
and for the
.Xr  man 7
.Ic TH
macro.
.It Fl T Ar output
Output format.
The
.Ar output
argument can be
.Cm ascii ,
.Cm utf8 ,
or
.Cm html ;
see
.Xr mandoc 1 .
In
.Cm html
output mode, the
.Cm fragment
output option is implied.
Other output options are not supported.
.It Fl v
Verbose mode, printing additional information to standard error output.
Specifying this once prints a summary about the number of files
and directories processed at the end of the iteration.
Specifying it twice additionally prints debugging information
about the backchannel from
.Xr mandocd 8
to
.Nm
that is used to limit the number of files in flight at any given time.
For details, see
.Sx DIAGNOSTICS .
.El
.Sh IMPLEMENTATION NOTES
Since this version avoids
.Xr fork 2
and
.Xr exec 3
overhead and uses the much faster
.Sy mandoc
parsers and formatters rather than
.Sy groff ,
it may be about one order of magnitude faster than other
.Nm  catman
implementations.
.Sh EXIT STATUS
.Ex -std
.Pp
Failures while trying to open individual manual pages for reading,
to save individual formatted files to the file system,
or even to read or create subdirectories do not cause
.Nm
to return an error exit status.
In such cases,
.Nm
simply continues with the next file or subdirectory.
.Sh DIAGNOSTICS
Some fatal errors cause
.Nm
to exit before the iteration over input files is even started:
.Bl -tag -width Ds -offset indent
.It unknown option \-\- Ar option
An invalid option was passed on the command line.
.It missing arguments: srcdir and dstdir
No argument was provided.
Both
.Ar srcdir
and
.Ar dstdir
are mandatory.
.It missing argument: dstdir
Only one argument was provided.
The second argument,
.Ar dstdir ,
is mandatory, too.
.It too many arguments: Ar third argument
Three or more arguments were provided, but only two are supported.
.It Sy socketpair : Ar reason
The sockets needed for communication with
.Xr mandocd 8
could not be created, for example due to file descriptor or memory exhaustion.
.It Sy fork : Ar reason
The new process needed to run
.Xr mandocd 8
could not be created, for example due to process table exhaustion
or system resource limits.
.It Sy exec Ns Po Sy mandocd Pc : Ar reason
The
.Xr mandocd 8
child program could not be started, for example because it is not in the
.Ev PATH
or has no execute permission.
.It Sy mkdir No destination Ar dstdir : reason
The
.Ar dstdir
does not exist and could not be created, for example because
the parent directory does not exist or permission is denied.
.It Sy open No destination Ar dstdir : reason
The
.Ar dstdir
could not be opened for reading, for example because
it is not a directory or permission is denied.
.It Sy chdir No to source Ar srcdir : reason
The current working directory could not be changed to
.Ar srcdir ,
for example because it does not exist, it is not a directory,
or permission is denied.
.It Sy fts_open : Ar reason
Starting the iteration was attempted but failed,
for example due to memory exhaustion.
.El
.Pp
Some fatal errors cause the iteration over input files to be aborted
prematurely:
.Bl -tag -width Ds -offset indent
.It FATAL: Sy fts_read : Ar reason
A call to
.Xr fts_read 3
returned
.Dv NULL ,
meaning that the iteration failed before being complete.
.It FATAL: mandocd child died: got Ar SIGNAME
This message appears if
.Nm
gets the
.Dv SIGCHLD
or
.Dv SIGPIPE
signal, most likely due to a fatal bug in
.Xr mandocd 8 .
.It FATAL: Sy sendmsg : Ar reason
The file descriptors needed to process one of the manual pages
could not be sent to
.Xr mandocd 8 ,
for example because
.Xr mandocd 8
could not be started or died unexpectedly.
.It FATAL: Sy recv : Ar reason
Trying to read a reply message from
.Xr mandocd 8
failed, most likely because
.Xr mandocd 8
unexpectedly died or closed the socket.
.It FATAL: signal Ar SIGNAME
This message appears if
.Nm
gets a
.Dv SIGHUP ,
.Dv SIGINT ,
or
.Dv SIGTERM
signal, for example because the user deliberately killed it.
.El
.Pp
Some non-fatal errors cause a single subdirectory to be skipped.
The iteration is not aborted but continues with the next subdirectory,
and the exit status is unaffected:
.Bl -tag -width Ds -offset indent
.It directory Ar subdirectory No unreadable : Ar reason
A directory below
.Ar srcdir
could not be read and is skipped.
.It directory Ar subdirectory No causes cycle
A directory below
.Ar srcdir
is skipped because it would cause cyclic processing.
.It Sy mkdirat Ar subdirectory : reason
A required directory below
.Ar dstdir
does not exist and could not be created.
The corresponding subdirectory below
.Ar srcdir
is skipped.
.El
.Pp
Some non-fatal errors cause a single source file to be skipped.
The iteration is not aborted but continues with the next file,
and the exit status is unaffected:
.Bl -tag -width Ds -offset indent
.It file Ar filename : reason
The function
.Xr fts_read 3
reported a non-fatal error with respect to
.Ar filename .
.It file Ar filename : No not a regular file
For example, it might be a symbolic link or a device file.
.It Sy open Ar filename No for reading : Ar reason
A file below
.Ar srcdir
could not be read, for example due to permission problems.
.It Sy openat Ar filename No for writing : Ar reason
A file below
.Ar dstdir
could not be created or truncated, for example due to permission problems.
.El
.Pp
If errors occur, the applicable summary messages appear
after the end of the iteration:
.Pp
.Bl -tag -width Ds -offset indent -compact
.It skipped Ar number No directories due to errors
.It skipped Ar number No files due to errors
.It processing aborted due to fatal error
.El
.Pp
If the
.Fl v
flag is specified, the following summary message also appears:
.Bl -tag -width Ds -offset indent
.It processed Ar nfiles No files in Ar ndirs No directories
A file is counted if it could be opened for reading and the
corresponding output file could be opened for writing;
this does not necessarily mean that it is a useful manual page.
A directory is counted if it could be opened for reading and the
corresponding output directory existed or could be created;
this does not necessarily mean that any files could be
processed inside.
.El
.Pp
If the
.Fl v
flag is specified twice, the following messages also appear:
.Bl -tag -width Ds -offset indent
.It allowing up to Ar number No files in flight
This is printed at the beginning of the iteration,
showing the maximum number of files that
.Nm
allows to be in flight at any given time.
.It files in flight: Ar old No \- Ar decrement No = Ar new
This message is printed when
.Nm
learns about
.Xr mandocd 8
accepting more than one file at the same time.
The three numbers printed are the old number of files in flight,
the amount this number is being reduced, and the resulting
new number of files in flight.
.It waiting for Ar number No files in flight
This message is printed at the end of the iteration, after
.Nm
has submitted all files to
.Xr mandocd 8
that it intends to.
THe message informs about the number of files still in flight
at this point.
The
.Nm
program then waits until
.Xr mandocd 8
has accepted them all or until an error occurs.
.El
.Sh SEE ALSO
.Xr mandoc 1 ,
.Xr mandocd 8
.Sh HISTORY
A
.Nm  catman
utility first appeared in
.Fx 1.0 .
Other, incompatible implementations appeared in
.Nx 1.0
and in
.Sy man-db No 2.2 .
.Pp
This version appeared in version 1.14.1 of the
.Sy mandoc
toolkit.
.Sh AUTHORS
.An -nosplit
The first
.Nm  catman
implementation was a short shell script by
.An Christoph Robitschko
in July 1993.
.Pp
The
.Nx
implementations were written by
.An J. T. Conklin Aq Mt jtc@netbsd.org
in 1993,
.An Christian E. Hopps Aq Mt chopps@netbsd.org
in 1994,
and
.An Dante Profeta Aq Mt dante@netbsd.org
in 1999; the
.Sy man-db
implementation by
.An Graeme W. Wilford
in 1994; and the
.Fx
implementations by
.An Wolfram Schneider Aq Mt wosch@freebsd.org
in 1995 and
.An John Rochester Aq Mt john@jrochester.org
in 2002.
.Pp
The concept of the present version was designed and implemented by
.An Michael Stapelberg Aq Mt stapelberg@debian.org
in 2017.
Option and argument handling and directory iteration was added by
.An Ingo Schwarze Aq Mt schwarze@openbsd.org .
.Sh CAVEATS
All versions of
.Nm  catman
are incompatible with each other because each caters to the needs
of a specific operating system, for example regarding directory
structures and file naming conventions.
.Pp
This version is more flexible than the others in so far as it does
not assume any particular directory structure or naming convention.
That flexibility comes at the price of not being able to change the
names and relative paths of the source files when reusing them to
store the formatted files, of not supporting any configuration file
formats or environment variables, and of being unable to scan for
and remove junk files in
.Ar dstdir .
.Pp
Currently,
.Nm
always reformats each page, even if the formatted version is newer
than the source version.


NEWS
$Id$

This file lists the most important changes in the mandoc.bsd.lv distribution.

Changes in the upcoming version 1.15.1, to be released in 2026

    --- MAJOR NEW FEATURES ---
 * man(1): use unveil(2) if provided by the operating system
 * mdoc(7): support .Lb libname [...] in the SYNOPSIS section
 * man(7): implement .MR (reference manual page), a groff-1.23 extension
    --- MAJOR INCOMPATIBLE CHANGE FOR GROFF-1.23 COMPATIBILITY ---
 * man(7)/tbl(7): stop unconditionally emitting vertical space before .TS
    --- MINOR NEW FEATURES ---
 * mdoc(7): support .St -isoC-2023 = ISO/IEC 9899:2024 ("ISO C23") (sic!)
 * mdoc(7): support .St -p1003.1-2024 = IEEE Std 1003.1-2024 ("POSIX.1")
 * mdoc(7) -T html: emit hyperlinks to rfc-editor.org for .%R RFC <number>
 * tbl(7): support numerical expressions in column width modifiers
 * roff(7): support decimal fractions in the number parser
 * roff(7): rudimentary implementation of \A (name syntax check)
 * roff(7): implement \(.l, but intentionally set it to the constant 78n
 * roff(7): implement \V, but intentionally do not inspect the environment
 * roff(7): implement .stringup and .stringdown introduced in groff-1.23
 * catman(8): new command line option -v (verbose mode)
    --- HTML ACCESSIBILITY IMPROVEMENTS ---
 * man.cgi(8): use <header>, <nav>, <footer>, and role & aria-label attributes
 * use flexbox CSS instead of the archaic <table> for headers and footers
 * use the <main> element, which has has the ARIA role "main" by default
 * use <h2> for .Sh/.SH and <h3> for .Ss/.SS, freeing <h1> for header.html
 * mdoc(7) .Nd: use the ARIA role "doc-subtitle"
 * mdoc(7) .Xr: print aria-label attributes on <a> elements
 * -O toc: print <nav role="doc-toc"> elements where appropriate
 * make sure <meta charset="utf-8"> is within the first 1024 bytes
    --- MINOR FEATURE REMOVALS ---
 * man(1): without -l, never interpret "name" command line arguments as
   absolute or relative path names, not even for arguments that contain
   a slash and that do not resolve to a page name; this was a quirk
   that even the maintainer of the man-db package dislikes
 * man(7): stop supporting the obsolete -O mdoc command line option,
   which no longer provided any functionality
    --- RELIABILITY BUGFIXES ---
 * mdoc(7) .Bl -column: avoid NULL pointer access with .EQ without .It
 * mdoc(7) .Tg: avoid segfault triggered by .Tg inside some blocks
 * mdoc(7) .Tg: avoid assert(3) on a node both implicitly and implicitly tagged
 * man(7): do not fail an assertion on sequences of multiple macros
   with next line scopes without arguments, like .SH .I .I .BI
 * tbl(7): fix an endless loop caused by certain expanding columns
 * tbl(7): do not read uninitialized memory and produce near-infinite output
   when the last data row of a table contains nothing but a horizontal line
 * tbl(7): do not read one byte past the end of an empty data line 
 * tbl(7): do not fail an assertion on \z near the end of a table cell
 * tbl(7): do not fail an assertion if ".if 1 .ce" occurs in a table
 * eqn(7): prevent an infinite loop caused by recursive "define"
 * roff(7): avoid a read buffer overrun if an escape sequence that requires
   an argument occurs at the end of an input line and the argument is missing
 * roff(7): do not segfault on .shift with a negative argument
 * roff(7): do not segfault on some macro args ending in double backslashes
 * roff(7): do not fail an assertion on \o inside .mc
 * roff(7): do not fail assertions on some lines containing two .while requests
 * roff(7): do not fail an assertion on expanding escapes starting with \EE
 * roff(7): do not fail an assertion on \zX\h'-N' at the beginning of a line
 * roff(7): do not fail assertions for certain nested escape sequences
    --- PORTABILITY IMPROVEMENTS ---
 * on systems that rename or relocate programs or manuals, patch the
   manuals before installation such that they use the correct names
 * configure: scaffolding to build without unveil(2) when unavailable
 * configure: support TEST_EXECUTION=0 for use in cross compilation
 * man.conf(5): make ETCDIR configurable, for systems unable to access /etc
 * support optionally building, installing, and using a libmandoc.so library
 * add missing <stddef.h> in strsep(3) compatibilty implementation
    --- MINOR FUNCTIONAL IMPROVEMENTS FOR GROFF-1.23 COMPATIBILITY ---
 * mdoc(7): change the footer line to always show NAME(SECTION) on the
   right hand side, for compatibility with groff-1.23 and with man(7)
 * mdoc(7) .Lk: set the URI in roman rather than in bold font
 * mdoc(7) .Ux: change rendering of this deprecated macro from "UNIX" to "Unix"
 * mdoc(7) .Dt: if the second argument (the section number) is invalid,
   print "LOCAL" for the volume title, just like when the argument is missing
 * man(7): reduce global indentation to 5n to match mdoc(7) rendering
 * man(7) .TH: if the first argument is empty, use the
   fixed string "UNTITLED" instead, just like in mdoc(7) does
 * man(7): ignore .br before the first .SH
 * man(7): for the purpose of vertical spacing, ignore .RS before a paragraph
 * tbl(7): separate a vertical line on the left side of a table by horizontal
   whitespace of 1n from the content of the leftmost column
 * roff(7): ignore various new requests that are irrelevant in manuals
 * terminal formatter: use bold italic for italic font esc in a section header
    --- OTHER MINOR FUNCTIONAL IMPROVEMENTS ---
 * man(7): warn about blank input lines just like in mdoc(7)
 * man(7) -T html: avoid <br/> inside <pre>, which caused ugly formatting
 * roff(7): support some practically relevant escape sequences in \w arguments
 * roff(7): more precise -W error messages about invalid escape sequences
 * terminal formatter: internally calculate many widths in BU rather than EN,
   to make rounding for horizontal spacing more compatible with groff
 * makewhatis(8): make ordering in mandoc.db(5) deterministic even in

catman.c
/* $Id$ */
/*
 * Copyright (c) 2017, 2025 Ingo Schwarze <schwarze@openbsd.org>
 * Copyright (c) 2017 Michael Stapelberg <stapelberg@debian.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "config.h"

#if NEED_XPG4_2
#define _XPG4_2
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>

#include <assert.h>
#if HAVE_ERR
#include <err.h>
#endif
#include <errno.h>
#include <fcntl.h>
#if HAVE_FTS
#include <fts.h>
#else
#include "compat_fts.h"
#endif
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

int		verbose_flag = 0;
sig_atomic_t	got_signal = 0;

int	 process_manpage(int, int, const char *);
int	 process_tree(int, int);
void	 run_mandocd(int, const char *, const char *)
		__attribute__((__noreturn__));
void	 signal_handler(int);
ssize_t	 sock_fd_write(int, int, int, int);
void	 usage(void) __attribute__((__noreturn__));


void
signal_handler(int signum)
{
	got_signal = signum;
}

void
run_mandocd(int sockfd, const char *outtype, const char* defos)
{
	char	 sockfdstr[10];
	int	 len;

	len = snprintf(sockfdstr, sizeof(sockfdstr), "%d", sockfd);
	if (len >= (int)sizeof(sockfdstr)) {
		errno = EOVERFLOW;
		len = -1;
	}
	if (len < 0)
		err(1, "snprintf");
	if (defos == NULL)
		execlp("mandocd", "mandocd", "-T", outtype,
		    sockfdstr, (char *)NULL);
	else
		execlp("mandocd", "mandocd", "-T", outtype,
		    "-I", defos, sockfdstr, (char *)NULL);
	err(1, "exec(mandocd)");
}

ssize_t
sock_fd_write(int fd, int fd0, int fd1, int fd2)
{
	const struct timespec timeout = { 0, 10000000 };  /* 0.01 s */
	struct msghdr	 msg;
	struct iovec	 iov;
	union {
		struct cmsghdr	 cmsghdr;
		char		 control[CMSG_SPACE(3 * sizeof(int))];
	} cmsgu;
	struct cmsghdr	*cmsg;
	int		*walk;
	ssize_t		 sz;
	unsigned char	 dummy[1] = {'\0'};

	iov.iov_base = dummy;
	iov.iov_len = sizeof(dummy);

	msg.msg_name = NULL;
	msg.msg_namelen = 0;
	msg.msg_iov = &iov;
	msg.msg_iovlen = 1;

	msg.msg_control = cmsgu.control;
	msg.msg_controllen = sizeof(cmsgu.control);

	cmsg = CMSG_FIRSTHDR(&msg);
	cmsg->cmsg_len = CMSG_LEN(3 * sizeof(int));
	cmsg->cmsg_level = SOL_SOCKET;
	cmsg->cmsg_type = SCM_RIGHTS;

	walk = (int *)CMSG_DATA(cmsg);
	*(walk++) = fd0;
	*(walk++) = fd1;
	*(walk++) = fd2;

	/*
	 * It appears that on some systems, sendmsg(3)
	 * may return EAGAIN even in blocking mode.
	 * Seen for example on Oracle Solaris 11.2.
	 * The sleeping time was chosen by experimentation,
	 * to neither cause more than a handful of retries
	 * in normal operation nor unnecessary delays.
	 */
	while ((sz = sendmsg(fd, &msg, 0)) == -1) {
		if (errno != EAGAIN) {
			warn("FATAL: sendmsg");
			break;
		}
		nanosleep(&timeout, NULL);
	}
	return sz;
}

int
process_manpage(int srv_fd, int dstdir_fd, const char *path)
{
	int	 in_fd, out_fd;
	int	 irc;

	if ((in_fd = open(path, O_RDONLY)) == -1) {
		warn("open %s for reading", path);
		fflush(stderr);
		return 0;
	}

	if ((out_fd = openat(dstdir_fd, path,
	    O_WRONLY | O_NOFOLLOW | O_CREAT | O_TRUNC,
	    S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) == -1) {
		warn("openat %s for writing", path);
		fflush(stderr);
		close(in_fd);
		return 0;
	}

	irc = sock_fd_write(srv_fd, in_fd, out_fd, STDERR_FILENO);

	close(in_fd);
	close(out_fd);

	return irc;
}

int
process_tree(int srv_fd, int dstdir_fd)
{
	const struct timespec timeout = { 0, 10000000 };  /* 0.01 s */
	const int	 max_inflight = 16;

	FTS		*ftsp;
	FTSENT		*entry;
	const char	*argv[2];
	const char	*path;
	int		 inflight, irc, decr, fatal;
	int		 gooddirs, baddirs, goodfiles, badfiles;
	char		 dummy[1];

	argv[0] = ".";
	argv[1] = (char *)NULL;

	if ((ftsp = fts_open((char * const *)argv,
	    FTS_PHYSICAL | FTS_NOCHDIR, NULL)) == NULL) {
		warn("fts_open");
		return -1;
	}

	if (verbose_flag >= 2) {
		warnx("allowing up to %d files in flight", max_inflight);
		fflush(stderr);
	}
	inflight = fatal = gooddirs = baddirs = goodfiles = badfiles = 0;
	while (fatal == 0 && got_signal == 0 &&
	    (entry = fts_read(ftsp)) != NULL) {
		if (inflight >= max_inflight) {
			while (recv(srv_fd, dummy, sizeof(dummy), 0) == -1) {
				if (errno != EAGAIN) {
					warn("FATAL: recv");
					fatal = errno;
					break;
				}
				nanosleep(&timeout, NULL);
			}
			if (fatal != 0)
				break;
			decr = 1;
			while ((irc = recv(srv_fd, dummy, sizeof(dummy),
			    MSG_DONTWAIT)) > 0)
				decr++;
			assert(inflight >= decr);
			if (verbose_flag >= 2 && decr > 1) {
				warnx("files in flight: %d - %d = %d",
				    inflight, decr, inflight - decr);
				fflush(stderr);
			}
			inflight -= decr;
			if (irc == 0) {
				errno = ECONNRESET;
				inflight = -1;
			}
			if (errno != EAGAIN) {
				warn("FATAL: recv");
				fatal = errno;
				break;
			}
		}
		path = entry->fts_path + 2;
		switch (entry->fts_info) {
		case FTS_F:
			switch (process_manpage(srv_fd, dstdir_fd, path)) {
			case -1:
				fatal = errno;
				break;
			case 0:
				badfiles++;
				break;
			default:
				goodfiles++;
				inflight++;
				break;
			}
			break;
		case FTS_D:
			if (*path != '\0' &&
			    mkdirat(dstdir_fd, path, S_IRWXU | S_IRGRP |
			      S_IXGRP | S_IROTH | S_IXOTH) == -1 &&
			    errno != EEXIST) {
				warn("mkdirat %s", path);
				fflush(stderr);
				(void)fts_set(ftsp, entry, FTS_SKIP);
				baddirs++;
			} else
				gooddirs++;
			break;
		case FTS_DP:
			break;
		case FTS_DNR:
			warnx("directory %s unreadable: %s",
			    path, strerror(entry->fts_errno));
			fflush(stderr);
			baddirs++;
			break;
		case FTS_DC:
			warnx("directory %s causes cycle", path);
			fflush(stderr);
			baddirs++;
			break;
		case FTS_ERR:
		case FTS_NS:
			warnx("file %s: %s",
			    path, strerror(entry->fts_errno));
			fflush(stderr);
			badfiles++;
			break;
		default:
			warnx("file %s: not a regular file", path);
			fflush(stderr);
			badfiles++;
			break;
		}
	}
	if (got_signal != 0) {
		switch (got_signal) {
		case SIGCHLD:
			warnx("FATAL: mandocd child died: got SIGCHLD");
			break;
		case SIGPIPE:
			warnx("FATAL: mandocd child died: got SIGPIPE");
			break;
		default:
			warnx("FATAL: signal SIG%s", sys_signame[got_signal]);
			break;
		}
		inflight = -1;
		fatal = 1;
	} else if (fatal == 0 && (fatal = errno) != 0)
		warn("FATAL: fts_read");

	fts_close(ftsp);
	if (verbose_flag >= 2 && inflight > 0) {
		warnx("waiting for %d files in flight", inflight);
		fflush(stderr);
	}
	while (inflight > 0) {
		irc = recv(srv_fd, dummy, sizeof(dummy), 0);
		if (irc > 0)
			inflight--;
		else if (irc == -1 && errno == EAGAIN)
			nanosleep(&timeout, NULL);
		else {
			if (irc == 0)
				errno = ECONNRESET;
			warn("recv");
			inflight = -1;
		}
	}
	if (verbose_flag)
		warnx("processed %d files in %d directories",
		    goodfiles, gooddirs);
	if (baddirs > 0)
		warnx("skipped %d %s due to errors", baddirs,
		    baddirs == 1 ? "directory" : "directories");
	if (badfiles > 0)
		warnx("skipped %d %s due to errors", badfiles,
		    badfiles == 1 ? "file" : "files");
	if (fatal != 0) {
		warnx("processing aborted due to fatal error, "
		    "results are probably incomplete");
		inflight = -1;
	}
	return inflight;
}

int
main(int argc, char **argv)
{
	struct sigaction sa;
	const char	*defos, *outtype;
	int		 srv_fds[2];
	int		 dstdir_fd;
	int		 opt;
	pid_t		 pid;

	defos = NULL;
	outtype = "ascii";
	while ((opt = getopt(argc, argv, "I:T:v")) != -1) {
		switch (opt) {
		case 'I':
			defos = optarg;
			break;
		case 'T':
			outtype = optarg;
			break;
		case 'v':
			verbose_flag += 1;
			break;
		default:
			usage();
		}
	}

	if (argc > 0) {
		argc -= optind;
		argv += optind;
	}
	if (argc != 2) {
		switch (argc) {
		case 0:
			warnx("missing arguments: srcdir and dstdir");
			break;
		case 1:
			warnx("missing argument: dstdir");
			break;
		default:
			warnx("too many arguments: %s", argv[2]);
			break;
		}
		usage();
	}

	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = &signal_handler;
	sa.sa_flags = SA_NOCLDWAIT;
	if (sigfillset(&sa.sa_mask) == -1)
		err(1, "sigfillset");
	if (sigaction(SIGHUP, &sa, NULL) == -1)
		err(1, "sigaction(SIGHUP)");
	if (sigaction(SIGINT, &sa, NULL) == -1)
		err(1, "sigaction(SIGINT)");
	if (sigaction(SIGPIPE, &sa, NULL) == -1)
		err(1, "sigaction(SIGPIPE)");
	if (sigaction(SIGTERM, &sa, NULL) == -1)
		err(1, "sigaction(SIGTERM)");
	if (sigaction(SIGCHLD, &sa, NULL) == -1)
		err(1, "sigaction(SIGCHLD)");

	if (socketpair(AF_LOCAL, SOCK_STREAM, AF_UNSPEC, srv_fds) == -1)
		err(1, "socketpair");

	pid = fork();
	switch (pid) {
	case -1:
		err(1, "fork");
	case 0:
		close(srv_fds[0]);
		run_mandocd(srv_fds[1], outtype, defos);
	default:
		break;
	}
	close(srv_fds[1]);

	if ((dstdir_fd = open(argv[1], O_RDONLY | O_DIRECTORY)) == -1) {
		if (errno != ENOENT)
			err(1, "open destination %s", argv[1]);
		if (mkdir(argv[1], S_IRWXU |
		    S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH) == -1) 
			err(1, "mkdir destination %s", argv[1]);
		if ((dstdir_fd = open(argv[1], O_RDONLY | O_DIRECTORY)) == -1)
			err(1, "open destination %s", argv[1]);
	}

	if (chdir(argv[0]) == -1)
		err(1, "chdir to source %s", argv[0]);

	return process_tree(srv_fds[0], dstdir_fd) == -1 ? 1 : 0;
}

void
usage(void)
{
	fprintf(stderr, "usage: %s [-I os=name] [-T output] "
	    "srcdir dstdir\n", BINM_CATMAN);
	exit(1);
}
